frontend/deb/debroot.go

```diff
@@ -0,0 +1,438 @@
+	"runtime/debug"
+	"sync"
+
+	"github.com/Azure/dalec"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/pkg/errors"
+)
+
+const customSystemdPostinstFile = "custom_systemd_postinst.sh.partial"
+
+//go:embed templates/patch-header.txt
+var patchHeader []byte
+
+// This creates a directory in the debian root directory for each patch, and copies the patch files into it.
+// The format for each patch dir matches what would normaly be under `debian/patches`, just that this is a separate dir for every source we are patching
+// This is purely for documenting in the source package how patches are applied in a more readable way than the big merged patch file.
+func sourcePatchesDir(sOpt dalec.SourceOpts, base llb.State, dir, name string, spec *dalec.Spec, opts ...llb.ConstraintsOpt) ([]llb.State, error) {
+	patchesPath := filepath.Join(dir, name)
+	base = base.
+		File(llb.Mkdir(patchesPath, 0o755), opts...)
+
+	var states []llb.State
+
+	seriesBuf := bytes.NewBuffer(nil)
+	for _, patch := range spec.Patches[name] {
+		src := spec.Sources[patch.Source]
+
+		copySrc := patch.Source
+		if patch.Path != "" {
+			src.Includes = append(src.Includes, patch.Path)
+			copySrc = filepath.Base(patch.Path)
+		}
+		st, err := src.AsState(patch.Source, sOpt, opts...)
+		if err != nil {
+			return nil, errors.Wrap(err, "error creating patch state")
+		}
+
+		st = base.File(llb.Copy(st, copySrc, filepath.Join(patchesPath, patch.Source)), opts...)
+		if _, err := seriesBuf.WriteString(name + "\n"); err != nil {
+			return nil, errors.Wrap(err, "error writing to series file")
+		}
+		states = append(states, st)
+	}
+
+	series := base.File(llb.Mkfile(filepath.Join(patchesPath, "series"), 0o640, seriesBuf.Bytes()), opts...)
+
+	return append(states, series), nil
+}
+
+// Debroot creates a debian root directory suitable for use with debbuild.
+// This does not include sources in case you want to mount sources (instead of copying them) later.
+func Debroot(sOpt dalec.SourceOpts, spec *dalec.Spec, worker, in llb.State, target, dir string, opts ...llb.ConstraintsOpt) (llb.State, error) {
+	control, err := controlFile(spec, in, target, dir)
+	if err != nil {
+		return llb.Scratch(), errors.Wrap(err, "error generating control file")
+	}
+
+	rules, err := Rules(spec, in, dir)
+	if err != nil {
+		return llb.Scratch(), errors.Wrap(err, "error generating rules file")
+	}
+
+	changelog, err := Changelog(spec, in, target, dir)
+	if err != nil {
+		return llb.Scratch(), errors.Wrap(err, "error generating changelog file")
+	}
+
+	if dir == "" {
+		dir = "debian"
+	}
+
+	base := llb.Scratch().File(llb.Mkdir(dir, 0o755), opts...)
+	installers := createInstallScripts(worker, spec, dir)
+
+	debian := base.
+		File(llb.Mkdir(filepath.Join(dir, "source"), 0o755), opts...).
+		With(func(in llb.State) llb.State {
+			if len(spec.Sources) == 0 {
+				return in
+			}
+			return in.
+				File(llb.Mkfile(filepath.Join(dir, "source/format"), 0o640, []byte("3.0 (quilt)")), opts...).
```

Comment:
Perhaps make the contents of the `source/format` and `source/options` files constants?